### ISSUE TITLE
Vivarium API Bugs & Enhancements

### ISSUE THREAD

**Issue by WilliamIX** [vivarium-collective/vivarium-core#86]
This issue will keep track of different bugs and enhancements in the comments as I test the Vivarium API by trying to create a set of swimming "proto-cell" agents on a lattice.

**Comment by WilliamIX** [comment]
Enhancement: Ability to generate Stores directly from Composers (instead of having to go through Composites)

**Comment by WilliamIX** [comment]
Bug: `Exception: trying to assign leaf values to a branch at: ('location',)` occurs when the substore 'location':{'x' : val, 'y':val} already exists.

Specifics:

```
S = store.Store({})
S["crn"] = Bioscrape(parameters)
S["motion"] = Motion({"time_step":1.0}) #A process with topology {'location':{'x':val, 'y':val}}
S['local_field'] = LocalField({})` #Also contains a topology {'location':{'x':val, 'y':val}}
```

This error message goes away if motion has the topology `{'loc':{'x':val, 'y':val}}` - I'm guessing it has to do with the schemas in LocalField not working correctly in the recursive call in `establish_path`

**Comment by WilliamIX** [comment]
Bug: adding a specific path to a star-schema:

This code:
`    lattice_store['agents', str(i), "local_field", 'exchanges', 'H'] = lattice_store['agents', str(i), 'crn', "delta_species", "H_external"]
`
Produces this error:
`Exception: topology for the process ('agents', '0', 'local_field') 
 at path ('agents', '0', 'exchanges') uses undeclared ports: {'H'}`

I believe because lattice_store['agents', str(i), "local_field", 'exchanges', 'H'] does not yet exist - I am trying to add a subpath to the path lattice_store['agents', str(i), "local_field", 'exchanges'] which targets one specific species.

**Comment by WilliamIX** [comment]
Bug: I can produce a Store which then breaks at runtime.

`lattice_composer = Lattice(lattice_config)
lattice_composite = lattice_composer.generate()
lattice_store = lattice_composite.generate_store()
...
for i in range(N_agents):
    lattice_store['agents', str(i), 'crn'] = Bioscrape(parameters)
    lattice_store['agents', str(i), 'motion'] = Motion({"time_step":1.0})    
    lattice_store['agents', str(i), 'local_field'] = LocalField({})
    lattice_store['agents', str(i), "motion", "delta", "H"] = lattice_store["agents", str(i), "crn", "delta_species", "H_internal"]
    lattice_store['agents', str(i), "local_field", 'fields'] = lattice_store['fields']`

Produces a store, which I can look at. However, when I try to simulate:
`sim = Engine({'store': lattice_store})

steps = 100
sim.update(steps)`

that store I get the error (in establish path):

`Exception: there is no path from leaf node ('agents', '0') to ('boundary',)
`

**Comment by WilliamIX** [comment]
Notes from meeting:
* To solve subschema issue: apply_subschema_path(path) in the setter of Store (do what we do in Store.add)

Mental Model:
* Set: sets the value of something that already exists (up to one level). Does not change topology.
* Get: gets the value of something that already exists
* Create Function: establishes a path and applies schema. Add should call create.
* Connect: throws an error unless you call it from a process node. Changes topology.

**Comment by WilliamIX** [comment]
For Dividing: make 'processes' and 'topology' optional keywords. If they don't exist, copy the existing topology and processes.

**Comment by eagmon** [comment]
Thanks for testing the API @WilliamIX! These issues are fixed with #87
